README.md
Raspi Board
===========

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/nebrius/raspi-io?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Build Status](https://travis-ci.org/nebrius/raspi-board.svg?branch=master)](https://travis-ci.org/nebrius/raspi-board)
[![Coverage Status](https://coveralls.io/repos/github/nebrius/raspi-board/badge.svg?branch=master)](https://coveralls.io/github/nebrius/raspi-board?branch=master)

Raspi Board provides pin mapping functionality for the various Raspberry Pi models. This module is meant to be run on a physical Raspberry Pi _only_, and maps pins for the specific board it is being run on.

If you have a bug report, feature request, or wish to contribute code, please be sure to check out the [Raspi IO Contributing Guide](https://github.com/nebrius/raspi-io/blob/master/CONTRIBUTING.md).

## System Requirements

- Raspberry Pi Model B Rev 1 or newer (sorry Model A users), except compute modules.
- Raspbian Jessie or newer
  - [Node-RED](http://nodered.org/) works, but can be finicky and difficult to debug.
  - See https://github.com/nebrius/raspi-io/issues/24 for more info about support for other OSes
- Node 6.0.0 or newer

Detailed instructions for getting a Raspberry Pi ready for NodeBots, including how to install Node.js, can be found in the [wiki](https://github.com/nebrius/raspi-io/wiki/Getting-a-Raspberry-Pi-ready-for-NodeBots)

## Installation

Install with npm:

```Shell
npm install raspi-board
```

**Note:** this project is written in [TypeScript](http://www.typescriptlang.org/) and includes type definitions in the package.json file. This means that if you want to use it from TypeScript, you don't need to install a separate @types module.

## Example

```JavaScript
const board = require('raspi-board');

// Prints 14
console.log(board.getPinNumber('SCLK0'));
```

## Pin Naming

The pins on the Raspberry Pi are a little complicated. There are multiple headers on some Raspberry Pis with extra pins, and the pin numbers are not consistent between Raspberry Pi board versions.

To help make it easier, you can specify pins in three ways. The first is to specify the pin by function, e.g. ```'GPIO18'```. The second way is to specify by pin number, which is specified in the form "P[header]-[pin]", e.g. ```'P1-7'```. The final way is specify the [Wiring Pi virtual pin number](http://wiringpi.com/pins/), e.g. ```7```. If you specify a number instead of a string, it is assumed to be a Wiring Pi number.

Be sure to read the [full list of pins](https://github.com/nebrius/raspi-io/wiki/Pin-Information) on the supported models of the Raspberry Pi.

## API

### Module Constants

<table>
  <thead>
    <tr>
      <th>Constant</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>VERSION_1_MODEL_A</td>
    <td>Constant representing the Raspberry Pi 1 Model A. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_1_MODEL_B_REV_1</td>
    <td>Constant representing the Raspberry Pi 1 Model B Rev 1. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_1_MODEL_B_REV_2</td>
    <td>Constant representing the Raspberry Pi 1 Model B Rev 1. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_1_MODEL_A_PLUS</td>
    <td>Constant representing the Raspberry Pi 1 Model A+. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_1_MODEL_B_PLUS</td>
    <td>Constant representing the Raspberry Pi 1 Model B+. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_1_MODEL_ZERO</td>
    <td>Constant representing the Raspberry Pi Zero. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_1_MODEL_ZERO_W</td>
    <td>Constant representing the Raspberry Pi Zero W. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_2_MODEL_B</td>
    <td>Constant representing the Raspberry Pi 2 Model B. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_3_MODEL_B</td>
    <td>Constant representing the Raspberry Pi 3 Model B. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_3_MODEL_B_PLUS</td>
    <td>Constant representing the Raspberry Pi 3 Model B+. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_3_MODEL_A_PLUS</td>
    <td>Constant representing the Raspberry Pi 3 Model A+. Returned from <code>getBoardRevision()</code></td>
  </tr>
  <tr>
    <td>VERSION_UNKNOWN</td>
    <td>Constant representing an unknown or unsupported version of the Raspberry Pi. Returned from <code>getBoardRevision()</code></td>
  </tr>
</table>

### getBoardRevision()

Gets the board revision.

_Arguments_: None.

_Returns_: A constant representing the board version, one of ```VERSION_1_MODEL_ZERO```, ```VERSION_1_MODEL_A```, ```VERSION_1_MODEL_B_REV_1```, ```VERSION_1_MODEL_B_REV_2```, ```VERSION_1_MODEL_B_PLUS```, ```VERSION_1_MODEL_A_PLUS```, ```VERSION_2_MODEL_B```, or ```VERSION_UNKNOWN```

### getPins()

Gets a list of all pins for the device it's running on.

_Arguments_: None.

_Returns_: An array of pin entry objects structured such that the array index is the Wiring Pi pin number, each with the following structure:

<table>
  <thead>
    <tr>
      <th>Property</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>pins</td>
    <td>Array</td>
    <td>An array of all the pin names for pin, e.g. <code>['GPIO21', 'SCLK1', 'P1-40']</code></td>
  </tr>
  <tr>
    <td>peripherals</td>
    <td>Array</td>
    <td>An array of all the peripherals supported on the pin, and may contain any combination of <code>'gpio'</code>, <code>'pwm'</code>, <code>'i2c'</code>, <code>'uart'</code>, or <code>'spi'</code></td>
  </tr>
  <tr>
    <td>gpio</td>
    <td>number</td>
    <td>The GPIO number for the pin, e.g. `4` for `GPIO4`. This method works regardless of pinMode</td>
  </tr>
</table>

### getPinNumber(pin)

Normalizes a pin name to a Wiring Pi pin number. This method accounts for the differences between board revisions by checking which Raspberry Pi the code is running on.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>pin</td>
    <td>String | Number</td>
    <td>The pin name/number to normalize. If a number is passed in, it is assumed to be the Wiring Pi pin number</td>
  </tr>
</table>

_Returns_: A number representing the Wiring Pi pin number, or null if an invalid pin name was passed in.

### getGpioNumber(pin)

Normalizes a pin name to a GPIO pin number, e.g. `TXD0`/`GPIO14` returns `14`. This method accounts for the differences between board revisions by checking which Raspberry Pi the code is running on.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>pin</td>
    <td>String | Number</td>
    <td>The pin name/number to normalize. If a number is passed in, it is assumed to be the Wiring Pi pin number</td>
  </tr>
</table>

_Returns_: A number representing the GPIO pin number, or null if an invalid pin name was passed in.

License
=======

The MIT License (MIT)

Copyright (c) 2014-2017 Bryan Hughes <bryan@nebri.us>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
## 7.3.3 (2020-1-17)

- Added another missing board

## 7.3.2 (2020-7-15)

- Added support for a handful of missing RPi models

## 7.3.1 (2020-3-9)

- Added support for another Raspberry Pi 4 Model B variant

## 7.3.0 (2019-9-12)

- Added better error checking so this module fails gracefully on non-Raspberry Pi systems (thanks to [@wolasss](https://github.com/wolasss) for the PR!)

## 7.2.3 (2019-9-03)

- Added support for another Raspberry Pi 4 Model B variant

## 7.2.2 (2019-7-17)

- Fixed a bug where the Raspberry Pi 4 wasn't properly wired in to the board detection logic

## 7.2.1 (2019-7-11)

- Added support for another Raspberry Pi 4 Model B variant

## 7.2.0 (2019-7-9)

- Added support for the Raspberry Pi 4 Model B

## 7.1.2 (2019-2-20)

- Updated dependencies

## 7.1.1 (2019-1-15)

- Fixed a dependency issue where "core-io-types" is not found

## 7.1.0 (2019-1-10)

- Reworked the exported pin info to use the Core IO Types pin info interfaces

## 7.0.0 (2019-1-9)

- BREAKING CHANGE: Removed `module` export as Core IO Types is also dropping a board specific module

## 6.2.0 (2018-12-5)

- Updated TypeScript to v3 (no compiled output change)
- Added `module` export to conform with Core IO types

## 6.1.0 (2018-11-29)

- Added support for the Raspberry Pi 3 Model A+

## 6.0.0 (2018-09-21)

- BREAKING CHANGE: dropped support for Node.js 4.x.x
- Rewrote unit tests to use Jasmine and bring it in line with other Raspi IO modules
- Rewrote test mode to use an environment variable instead of a global variable
- Upgraded dependencies, including bumping TypeScript to v3 and tweaking the TS configuration to target newer versions of JavaScript

## 5.3.1 (2018-06-12)

- Added missing RPi 3 Model B revision number

## 5.3.0 (2018-04-28)

- Added a missing RPi 1 Model B+ and RPi 3 Model B revision numbers
- Changed the "unknown board" warning message to an info message and added a much longer description.

## 5.2.0 (2018-04-21)

- Added support for the Raspberry Pi 3 Model B+

## 5.1.0 (2018-01-18)

- Added support for overclocked Pi Zero W's

## 5.0.2 (2017-11-18)

- Updated the tsconfig.json and tslint.json configuration files to use a newer style, build tools updated
- Cleaned up some TypeScript type definitions to be more compatible with DefinitelyTyped
- Note: there is no functionality change with this update

## 5.0.1 (2017-9-11)

- Non-code change: updated the README and republishing to make sure it's in npm

## 5.0.0 (2017-9-11)

- BREAKING CHANGE: dropped support for Node.js < 4.0.0, and now enforce it via package.json "engines" field.
- SORT OF BREAKING-ISH CHANGE: dropped support for attempting to install on non-arm platforms via package.json's "cpu" field.
    - Attempting to install this on a non-Raspberry Pi platform previously wasn't very useful because it usually crashed with a cryptic error. Now it complains earlier and harder.

## 4.2.0 (2017-4-21)

- Added the `getGpioNumber` method
- Added the `gpio` property on pin info returned from the `getPins` method
- Fixed some bugs around `VERSION_UNKNOWN` not being exported nor returned from `getBoardRevision`

## 4.1.0 (2017-4-19)

- Added support for the Raspberry Pi Zero W

## 4.0.2 (2017-1-22)

- Publishing a new version to update the README on npmjs.com. No other changes.

## 4.0.1 (2016-12-3)

- Converted the project to TypeScript and cleaned up a bunch of odds and ends
  - Note: there is no functionality change or bug fixes with this release

## 4.0.0 (2016-10-13)

- Added missing revision IDs for certain batches of the Raspbery Pi Zero
- Updated logic so an unknown board assumes Raspberry Pi 3 Model B pinout
  - POTENTIALLY BREAKING CHANGE: previous behavior was to throw an exception

## 3.2.0 (2016-9-19)

- Added missing revision ID for certain batches of the Raspberry Pi Model A+

## 3.1.0 (2016-6-21)

- Added missing revision ID for certain batches of the Raspbery Pi Zero

## 3.0.0 (2016-3-20)

- Fixed a bug where `PWM1` wasn't mapped to 23.
  - POTENTIALLY BREAKING CHANGE: using the identifier `PWM1` now maps to Wiring Pi pin 23, not 24
- New build system

## 2.4.1 (2016-3-7)

- Added a missing revision ID for certain batches of Raspberry Pi 3s

## 2.4.0 (2016-3-4)

- Added Raspberry Pi 3 Model B support

## 2.3.0 (2015-12-8)

- Added Raspberry Pi Zero support
- Refactored unit tests to use nodeunit

## 2.2.2 (2015-9-3)


src/index.ts
/*
The MIT License (MIT)

Copyright (c) 2014-2017 Bryan Hughes <bryan@nebri.us>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

import { readFileSync } from 'fs';
import { IPinInfo, PeripheralType } from 'j5-io-types';

export const VERSION_1_MODEL_A = 'rpi1_a';
export const VERSION_1_MODEL_B_REV_1 = 'rpi1_b1';
export const VERSION_1_MODEL_B_REV_2 = 'rpi1_b2';
export const VERSION_1_MODEL_B_PLUS = 'rpi1_bplus';
export const VERSION_1_MODEL_A_PLUS = 'rpi1_aplus';
export const VERSION_1_MODEL_ZERO = 'rpi1_zero';
export const VERSION_1_MODEL_ZERO_W = 'rpi1_zerow';
export const VERSION_2_MODEL_B = 'rpi2_b';
export const VERSION_3_MODEL_B = 'rpi3_b';
export const VERSION_3_MODEL_B_PLUS = 'rpi3_bplus';
export const VERSION_3_MODEL_A_PLUS = 'rpi3_aplus';
export const VERSION_4_MODEL_B = 'rpi4_b';
export const VERSION_UNKNOWN = 'unknown';

export interface IRaspiPinInfo extends IPinInfo {
  gpio: number;
}

const BOARD_REVISIONS: { [ revision: string ]: string } = {
  '0002': VERSION_1_MODEL_B_REV_1,
  '0003': VERSION_1_MODEL_B_REV_1,
  '0004': VERSION_1_MODEL_B_REV_2,
  '0005': VERSION_1_MODEL_B_REV_2,
  '0006': VERSION_1_MODEL_B_REV_2,
  '0007': VERSION_1_MODEL_A,
  '0008': VERSION_1_MODEL_A,
  '0009': VERSION_1_MODEL_A,
  '000d': VERSION_1_MODEL_B_REV_2,
  '000e': VERSION_1_MODEL_B_REV_2,
  '000f': VERSION_1_MODEL_B_REV_2,
  '0010': VERSION_1_MODEL_B_PLUS,
  '0012': VERSION_1_MODEL_A_PLUS,
  '0013': VERSION_1_MODEL_B_PLUS,
  '0015': VERSION_1_MODEL_A_PLUS,
  '900021': VERSION_1_MODEL_A_PLUS,
  '900032': VERSION_1_MODEL_B_PLUS,
  '900092': VERSION_1_MODEL_ZERO,
  '920092': VERSION_1_MODEL_ZERO,
  '900093': VERSION_1_MODEL_ZERO,
  '920093': VERSION_1_MODEL_ZERO,
  '9000c1': VERSION_1_MODEL_ZERO_W,
  '19000c1': VERSION_1_MODEL_ZERO_W, // Overclocked Pi Zero W
  'a01040': VERSION_2_MODEL_B,
  'a01041': VERSION_2_MODEL_B,
  'a21041': VERSION_2_MODEL_B,
  '2a21041': VERSION_2_MODEL_B,
  'a22042': VERSION_2_MODEL_B,
  'a02082': VERSION_3_MODEL_B,
  'a22082': VERSION_3_MODEL_B,
  'a32082': VERSION_3_MODEL_B,
  'a52082': VERSION_3_MODEL_B,
  '2a02082': VERSION_3_MODEL_B,
  'a020d3': VERSION_3_MODEL_B_PLUS,
  '2a020d3': VERSION_3_MODEL_B_PLUS,
  '9020e0': VERSION_3_MODEL_A_PLUS,
  'a03111': VERSION_4_MODEL_B, // 1GB RAM
  'b03111': VERSION_4_MODEL_B, // 2GB RAM
  'b03112': VERSION_4_MODEL_B,
  'c03111': VERSION_4_MODEL_B,  // 4GB RAM
  'c03112': VERSION_4_MODEL_B, // 4GB RAM
  'd03114': VERSION_4_MODEL_B // 8GB RAM
};

const B1: { [ wiringpi: number ]: IRaspiPinInfo } = {
  0: {
    pins: [
      'GPIO17',
      'P1-11'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 17
  },
  1: {
    pins: [
      'GPIO18',
      'PWM0',
      'P1-12'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.PWM
    ],
    gpio: 18
  },
  2: {
    pins: [
      'GPIO21',
      'P1-13'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 21
  },
  3: {
    pins: [
      'GPIO22',
      'P1-15'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 22
  },
  4: {
    pins: [
      'GPIO23',
      'P1-16'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 23
  },
  5: {
    pins: [
      'GPIO24',
      'P1-18'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 24
  },
  6: {
    pins: [
      'GPIO25',
      'P1-22'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 25
  },
  7: {
    pins: [
      'GPIO4',
      'P1-7'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 4
  },
  8: {
    pins: [
      'GPIO0',
      'SDA0',
      'P1-3'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.I2C
    ],
    gpio: 0
  },
  9: {
    pins: [
      'GPIO1',
      'SCL0',
      'P1-5'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.I2C
    ],
    gpio: 1
  },
  10: {
    pins: [
      'GPIO8',
      'P1-24',
      'CE0'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 8
  },
  11: {
    pins: [
      'GPIO7',
      'P1-26'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 7
  },
  12: {
    pins: [
      'GPIO10',
      'MOSI0',
      'P1-19'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 10
  },
  13: {
    pins: [
      'GPIO9',
      'MISO0',
      'P1-21'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 9
  },
  14: {
    pins: [
      'GPIO11',
      'SCLK0',
      'P1-23'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 11
  },
  15: {
    pins: [
      'GPIO14',
      'TXD0',
      'P1-8'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.UART
    ],
    gpio: 14
  },
  16: {
    pins: [
      'GPIO15',
      'RXD0',
      'P1-10'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.UART
    ],
    gpio: 15
  }
};

const B2: { [ wiringpi: number ]: IRaspiPinInfo } = {
  0: {
    pins: [
      'GPIO17',
      'P1-11'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 17
  },
  1: {
    pins: [
      'GPIO18',
      'PWM0',
      'P1-12'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.PWM
    ],
    gpio: 18
  },
  2: {
    pins: [
      'GPIO27',
      'P1-13'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 27
  },
  3: {
    pins: [
      'GPIO22',
      'P1-15'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 22
  },
  4: {
    pins: [
      'GPIO23',
      'P1-16'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 23
  },
  5: {
    pins: [
      'GPIO24',
      'P1-18'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 24
  },
  6: {
    pins: [
      'GPIO25',
      'P1-22'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 25
  },
  7: {
    pins: [
      'GPIO4',
      'P1-7'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 4
  },
  8: {
    pins: [
      'GPIO2',
      'SDA0',
      'P1-3'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.I2C
    ],
    gpio: 2
  },
  9: {
    pins: [
      'GPIO3',
      'SCL0',
      'P1-5'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.I2C
    ],
    gpio: 3
  },
  10: {
    pins: [
      'GPIO8',
      'CE0',
      'P1-24'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 8
  },
  11: {
    pins: [
      'GPIO7',
      'P1-26'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 7
  },
  12: {
    pins: [
      'GPIO10',
      'MOSI0',
      'P1-19'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 10
  },
  13: {
    pins: [
      'GPIO9',
      'MISO0',
      'P1-21'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 9
  },
  14: {
    pins: [
      'GPIO11',
      'SCLK0',
      'P1-23'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 11
  },
  15: {
    pins: [
      'GPIO14',
      'TXD0',
      'P1-8'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.UART
    ],
    gpio: 14
  },
  16: {
    pins: [
      'GPIO15',
      'RXD0',
      'P1-10'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.UART
    ],
    gpio: 15
  },
  17: {
    pins: [
      'GPIO28',
      'P5-3'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 28
  },
  18: {
    pins: [
      'GPIO29',
      'P5-4'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 29
  },
  19: {
    pins: [
      'GPIO30',
      'P5-5'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 30
  },
  20: {
    pins: [
      'GPIO31',
      'P5-6'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 31
  }
};

const BPLUS: { [ wiringpi: number ]: IRaspiPinInfo } = {
  0: {
    pins: [
      'GPIO17',
      'P1-11'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 17
  },
  1: {
    pins: [
      'GPIO18',
      'PWM0',
      'P1-12'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.PWM
    ],
    gpio: 18
  },
  2: {
    pins: [
      'GPIO27',
      'P1-13'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 27
  },
  3: {
    pins: [
      'GPIO22',
      'P1-15'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 22
  },
  4: {
    pins: [
      'GPIO23',
      'P1-16'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 23
  },
  5: {
    pins: [
      'GPIO24',
      'P1-18'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 24
  },
  6: {
    pins: [
      'GPIO25',
      'P1-22'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 25
  },
  7: {
    pins: [
      'GPIO4',
      'P1-7'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 4
  },
  8: {
    pins: [
      'GPIO2',
      'SDA0',
      'P1-3'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.I2C
    ],
    gpio: 2
  },
  9: {
    pins: [
      'GPIO3',
      'SCL0',
      'P1-5'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.I2C
    ],
    gpio: 3
  },
  10: {
    pins: [
      'GPIO8',
      'CE0',
      'P1-24'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 8
  },
  11: {
    pins: [
      'GPIO7',
      'CE1',
      'P1-26'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 7
  },
  12: {
    pins: [
      'GPIO10',
      'MOSI0',
      'P1-19'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 10
  },
  13: {
    pins: [
      'GPIO9',
      'MISO0',
      'P1-21'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 9
  },
  14: {
    pins: [
      'GPIO11',
      'SCLK0',
      'P1-23'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 11
  },
  15: {
    pins: [
      'GPIO14',
      'TXD0',
      'P1-8'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.UART
    ],
    gpio: 14
  },
  16: {
    pins: [
      'GPIO15',
      'RXD0',
      'P1-10'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.UART
    ],
    gpio: 15
  },
  21: {
    pins: [
      'GPIO5',
      'P1-29'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 5
  },
  22: {
    pins: [
      'GPIO6',
      'P1-31'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 6
  },
  23: {
    pins: [
      'GPIO13',
      'P1-33',
      'PWM1'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.PWM
    ],
    gpio: 13
  },
  24: {
    pins: [
      'GPIO19',
      'PWM1',
      'MISO1',
      'P1-35'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.PWM,
      PeripheralType.SPI
    ],
    gpio: 19
  },
  25: {
    pins: [
      'GPIO26',
      'P1-37'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 26
  },
  26: {
    pins: [
      'GPIO12',
      'PWM0',
      'P1-32'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.PWM
    ],
    gpio: 12
  },
  27: {
    pins: [
      'GPIO16',
      'P1-36'
    ],
    peripherals: [
      PeripheralType.GPIO
    ],
    gpio: 16
  },
  28: {
    pins: [
      'GPIO20',
      'MOSI1',
      'P1-38'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 20
  },
  29: {
    pins: [
      'GPIO21',
      'SCLK1',
      'P1-40'
    ],
    peripherals: [
      PeripheralType.GPIO,
      PeripheralType.SPI
    ],
    gpio: 21
  }
};

// Initialize the board info
let procInfo = '';
let rev = '';

if (process.env.RASPI_IO_TEST_MODE) {
  rev = 'a21041';
  procInfo = `Revision:${rev}`;
} else {
    try {
      procInfo = readFileSync('/proc/cpuinfo').toString();
      const revMatch = procInfo.match(/Revision\s*:\s*(.*)/);

      if (!revMatch) {
          console.warn('Unable to parse revision information in /proc/cpuinfo');
      } else {
          rev = revMatch[1];
      }
    } catch (e) {
      console.warn(`Unable to read /proc/cpuinfo file. Are you running this code on a Raspberry Pi? Error: ${e}`);
    }
}

// If the board has been overclocked, the revision is modified, so clear it here
if (/10[0-9a-z]{5}/.test(rev)) { // Check for RPi 1 overclock
  rev = rev.substr(-4);
} else if (/1a[0-9a-z]{5}/.test(rev)) { // Check for RPi 2 overclock
  rev = rev.substr(-6);
}

let pins: { [ wiringpi: number ]: IRaspiPinInfo };
switch (BOARD_REVISIONS[rev]) {
  case VERSION_1_MODEL_A:
    // Information is scarce, and no one has complained about it not being supported
    throw new Error('Raspberry Pi 1 Model A boards are not supported.');
  case VERSION_1_MODEL_B_REV_1:
    pins = B1;
    break;
  case VERSION_1_MODEL_B_REV_2:
    pins = B2;
    break;
  case VERSION_1_MODEL_ZERO:
  case VERSION_1_MODEL_ZERO_W:
  case VERSION_1_MODEL_A_PLUS:
  case VERSION_1_MODEL_B_PLUS:
  case VERSION_2_MODEL_B:
  case VERSION_3_MODEL_B:
  case VERSION_3_MODEL_B_PLUS:
  case VERSION_3_MODEL_A_PLUS:
  case VERSION_4_MODEL_B:
    pins = BPLUS;
    break;
  default:
    console.info(`Unknown board revision ${rev}, assuming Raspberry Pi Zero/2/3 pinout. ` +
      `Unless you are running a compute module or very old Raspberry Pi, you can safely ignore this message. ` +
      `Please report this board revision in a GitHub issue at https://github.com/nebrius/raspi-board.`);
    pins = BPLUS;
    break;
}

// Create the aliases
const aliases: { [ alias: string ]: number } = {};
for (const pin in pins) {
  if (pins.hasOwnProperty(pin)) {
    const pinAliases = pins[pin].pins;
    for (const pinAlias of pinAliases) {
      aliases[pinAlias] = parseInt(pin, 10);
    }
  }
}

// Create the Wiring Pi to PIGPIO mapping
const pigpioMapping: { [wiringPi: number]: number } = {};
for (const pin in pins) {
  if (pins.hasOwnProperty(pin)) {
    pigpioMapping[pin] = pins[pin].gpio;
  }
}

export function getBoardRevision() {
  return BOARD_REVISIONS[rev] || VERSION_UNKNOWN;
}

export function getPins() {
  return pins;
}

export function getPinNumber(alias: string | number): number | null {
  if (typeof alias !== 'number' && typeof alias !== 'string') {
    return null;
  }
  alias = alias.toString();
  if (Object.keys(pins).indexOf(alias) !== -1) {
    alias = parseInt(alias, 10);
  } else {
    alias = aliases[alias];
  }
  if (typeof alias === 'undefined') {
    return null;
  }
  return alias;
}

export function getGpioNumber(alias: string | number): number | null {
  const wiringpi = getPinNumber(alias);
  if (wiringpi === null) {
    return null;
  }
  return pigpioMapping[wiringpi];
}
